File: sdk/go/observability.go
package lelu

import (
	"context"
	"fmt"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/codes"
	"go.opentelemetry.io/otel/trace"
)

// AI Agent Semantic Conventions (matching backend implementation)
const (
	// Agent identification
	AttrAgentID       = "ai.agent.id"
	AttrAgentType     = "ai.agent.type"
	AttrAgentVersion  = "ai.agent.version"
	
	// Request attributes
	AttrRequestIntent     = "ai.request.intent"
	AttrRequestConfidence = "ai.request.confidence"
	AttrRequestActingFor  = "ai.request.acting_for"
	AttrRequestScope      = "ai.request.scope"
	
	// Policy evaluation
	AttrPolicyName    = "ai.policy.name"
	AttrPolicyVersion = "ai.policy.version"
	AttrPolicyResult  = "ai.policy.result"
	
	// Decision attributes
	AttrDecisionType       = "ai.decision.type"
	AttrDecisionConfidence = "ai.decision.confidence"
	AttrDecisionHumanReview = "ai.decision.human_review"
	AttrDecisionRiskScore  = "ai.decision.risk_score"
	AttrDecisionOutcome    = "ai.decision.outcome"
	
	// Multi-agent correlation
	AttrSwarmID         = "ai.swarm.id"
	AttrDelegationChain = "ai.delegation.chain"
	AttrParentAgent     = "ai.parent.agent"
	AttrChildAgent      = "ai.child.agent"
	
	// Performance metrics
	AttrLatencyMs        = "ai.latency.ms"
	AttrConfidenceGateMs = "ai.latency.confidence_gate_ms"
	AttrPolicyEvalMs     = "ai.latency.policy_eval_ms"
	AttrRiskEvalMs       = "ai.latency.risk_eval_ms"
)

// Agent types
const (
	AgentTypeAutonomous = "autonomous"
	AgentTypeAssisted   = "assisted"
	AgentTypeHuman      = "human"
)

// Decision types
const (
	DecisionTypeAutonomous = "autonomous"
	DecisionTypeHumanReview = "human_review"
	DecisionTypeDenied     = "denied"
)

// AgentTracer provides AI agent-aware tracing capabilities for the Go SDK
type AgentTracer struct {
	tracer trace.Tracer
}

// NewAgentTracer creates a new agent-aware tracer
func NewAgentTracer(serviceName string) *AgentTracer {
	return &AgentTracer{
		tracer: otel.Tracer(serviceName),
	}
}

// StartAgentSpan starts a new span with AI agent semantic conventions
func (t *AgentTracer) StartAgentSpan(ctx context.Context, operationName string, agentID string, attrs ...attribute.KeyValue) (context.Context, trace.Span) {
	// Add agent identification attributes
	agentAttrs := []attribute.KeyValue{
		attribute.String(AttrAgentID, agentID),
		attribute.String(AttrAgentType, AgentTypeAutonomous),
	}
	agentAttrs = append(agentAttrs, attrs...)
	
	ctx, span := t.tracer.Start(ctx, operationName)
	span.SetAttributes(agentAttrs...)
	
	return ctx, span
}

// StartAuthorizationSpan starts a span for agent authorization with full context
func (t *AgentTracer) StartAuthorizationSpan(ctx context.Context, agentID, action string, confidence float64) (context.Context, trace.Span) {
	return t.StartAgentSpan(ctx, "ai.agent.authorize",
		agentID,
		attribute.String(AttrRequestIntent, action),
		attribute.Float64(AttrRequestConfidence, confidence),
		attribute.String(AttrDecisionType, DecisionTypeAutonomous),
	)
}

// StartDelegationSpan starts a span for multi-agent delegation
func (t *AgentTracer) StartDelegationSpan(ctx context.Context, delegator, delegatee string) (context.Context, trace.Span) {
	return t.StartAgentSpan(ctx, "ai.agent.delegate",
		delegator,
		attribute.String(AttrParentAgent, delegator),
		attribute.String(AttrChildAgent, delegatee),
		attribute.String(AttrDelegationChain, fmt.Sprintf("%s→%s", delegator, delegatee)),
	)
}

// StartSwarmSpan starts a span for swarm orchestration
func (t *AgentTracer) StartSwarmSpan(ctx context.Context, swarmID string, orchestrator string, agents []string) (context.Context, trace.Span) {
	return t.StartAgentSpan(ctx, "ai.swarm.orchestration",
		orchestrator,
		attribute.String(AttrAgentType, "orchestrator"),
		attribute.String(AttrSwarmID, swarmID),
		attribute.StringSlice("ai.swarm.agents", agents),
		attribute.Int("ai.swarm.agent_count", len(agents)),
	)
}

// RecordPolicyEvaluation records policy evaluation details in the span
func (t *AgentTracer) RecordPolicyEvaluation(span trace.Span, policyName, policyVersion, result string, latencyMs float64) {
	if span == nil {
		return
	}
	span.SetAttributes(
		attribute.String(AttrPolicyName, policyName),
		attribute.String(AttrPolicyVersion, policyVersion),
		attribute.String(AttrPolicyResult, result),
		attribute.Float64(AttrPolicyEvalMs, latencyMs),
	)
}

// RecordDecision records the final authorization decision
func (t *AgentTracer) RecordDecision(span trace.Span, allowed bool, requiresHumanReview bool, confidence, riskScore float64, outcome string) {
	if span == nil {
		return
	}
	
	decisionType := DecisionTypeAutonomous
	if requiresHumanReview {
		decisionType = DecisionTypeHumanReview
	} else if !allowed {
		decisionType = DecisionTypeDenied
	}
	
	span.SetAttributes(
		attribute.String(AttrDecisionType, decisionType),
		attribute.Float64(AttrDecisionConfidence, confidence),
		attribute.Bool(AttrDecisionHumanReview, requiresHumanReview),
		attribute.Float64(AttrDecisionRiskScore, riskScore),
		attribute.String(AttrDecisionOutcome, outcome),
	)
	
	// Set span status based on decision
	if !allowed && !requiresHumanReview {
		span.SetStatus(codes.Error, "Authorization denied")
	} else {
		span.SetStatus(codes.Ok, "")
	}
}

// RecordLatency records operation latency
func (t *AgentTracer) RecordLatency(span trace.Span, totalMs, confidenceGateMs, policyEvalMs, riskEvalMs float64) {
	if span == nil {
		return
	}
	span.SetAttributes(
		attribute.Float64(AttrLatencyMs, totalMs),
		attribute.Float64(AttrConfidenceGateMs, confidenceGateMs),
		attribute.Float64(AttrPolicyEvalMs, policyEvalMs),
		attribute.Float64(AttrRiskEvalMs, riskEvalMs),
	)
}

// InjectCorrelationContext injects correlation context for multi-agent tracing
func (t *AgentTracer) InjectCorrelationContext(span trace.Span, delegationChain string) {
	if span == nil {
		return
	}
	
	span.SetAttributes(
		attribute.String(AttrDelegationChain, delegationChain),
	)
}

// WithAgentSpan executes a function within an agent span context
func (t *AgentTracer) WithAgentSpan(ctx context.Context, operationName string, agentID string, fn func(context.Context, trace.Span) error, attrs ...attribute.KeyValue) error {
	ctx, span := t.StartAgentSpan(ctx, operationName, agentID, attrs...)
	defer span.End()
	
	if err := fn(ctx, span); err != nil {
		span.RecordError(err)
		span.SetStatus(codes.Error, err.Error())
		return err
	}
	
	span.SetStatus(codes.Ok, "")
	return nil
}

// DecisionMetrics holds decision-related metrics for recording
type DecisionMetrics struct {
	Allowed             bool
	RequiresHumanReview bool
	Confidence          float64
	RiskScore           float64
	Outcome             string
}

// LatencyMetrics holds latency-related metrics for recording
type LatencyMetrics struct {
	TotalMs          float64
	ConfidenceGateMs float64
	PolicyEvalMs     float64
	RiskEvalMs       float64
}

// Default agent tracer instance
var defaultAgentTracer = NewAgentTracer("lelu-go-sdk")

// GetAgentTracer returns the default agent tracer instance
func GetAgentTracer() *AgentTracer {
	return defaultAgentTracer
}

// SetAgentTracer sets a custom agent tracer instance
func SetAgentTracer(tracer *AgentTracer) {
	defaultAgentTracer = tracer
}

File: sdk/python/auth_pe/observability.py
"""
Enhanced observability for AI agents - Phase 1 implementation.
Provides OpenTelemetry-based tracing with AI agent semantic conventions.
"""

import time
from typing import Dict, List, Optional, Any, Union, Tuple
from contextlib import contextmanager

try:
    from opentelemetry import trace as otel_trace
    from opentelemetry.trace import Status as OtelStatus, StatusCode as OtelStatusCode
    OTEL_AVAILABLE = True
except ImportError:
    OTEL_AVAILABLE = False
    # Provide stub classes for when OpenTelemetry is not available
    class OtelTrace:
        @staticmethod
        def get_tracer(name: str, version: Optional[str] = None) -> 'NoOpTracer':
            return NoOpTracer()
    
    class _StubStatus:
        pass
    
    class _StubStatusCode:
        OK = "OK"
        ERROR = "ERROR"
    
    otel_trace = OtelTrace()
    OtelStatus = _StubStatus
    OtelStatusCode = _StubStatusCode


class NoOpSpan:
    """No-op span implementation when OpenTelemetry is not available."""
    
    def set_attributes(self, attributes: Dict[str, Any]) -> None:
        pass
    
    def set_status(self, status: Union[OtelStatus, OtelStatusCode], description: Optional[str] = None) -> None:
        pass
    
    def record_exception(self, exception: Exception) -> None:
        pass
    
    def end(self) -> None:
        pass
    
    def __enter__(self) -> 'NoOpSpan':
        return self
    
    def __exit__(self, exc_type: Any, exc_val: Any, exc_tb: Any) -> None:
        pass


class NoOpTracer:
    """No-op tracer implementation when OpenTelemetry is not available."""
    
    def start_span(self, name: str, **kwargs: Any) -> NoOpSpan:
        return NoOpSpan()


# AI Agent Semantic Conventions (matching Go and TypeScript implementations)
class AIAgentAttributes:
    """AI Agent semantic conventions for OpenTelemetry."""
    
    # Agent identification
    AGENT_ID = "ai.agent.id"
    AGENT_TYPE = "ai.agent.type"
    AGENT_VERSION = "ai.agent.version"
    
    # Request attributes
    REQUEST_INTENT = "ai.request.intent"
    REQUEST_CONFIDENCE = "ai.request.confidence"
    REQUEST_ACTING_FOR = "ai.request.acting_for"
    REQUEST_SCOPE = "ai.request.scope"
    
    # Policy evaluation
    POLICY_NAME = "ai.policy.name"
    POLICY_VERSION = "ai.policy.version"
    POLICY_RESULT = "ai.policy.result"
    
    # Decision attributes
    DECISION_TYPE = "ai.decision.type"
    DECISION_CONFIDENCE = "ai.decision.confidence"
    DECISION_HUMAN_REVIEW = "ai.decision.human_review"
    DECISION_RISK_SCORE = "ai.decision.risk_score"
    DECISION_OUTCOME = "ai.decision.outcome"
    
    # Multi-agent correlation
    SWARM_ID = "ai.swarm.id"
    DELEGATION_CHAIN = "ai.delegation.chain"
    PARENT_AGENT = "ai.parent.agent"
    CHILD_AGENT = "ai.child.agent"
    
    # Performance metrics
    LATENCY_MS = "ai.latency.ms"
    CONFIDENCE_GATE_MS = "ai.latency.confidence_gate_ms"
    POLICY_EVAL_MS = "ai.latency.policy_eval_ms"
    RISK_EVAL_MS = "ai.latency.risk_eval_ms"


class AgentTypes:
    """Agent type constants."""
    AUTONOMOUS = "autonomous"
    ASSISTED = "assisted"
    HUMAN = "human"


class DecisionTypes:
    """Decision type constants."""
    AUTONOMOUS = "autonomous"
    HUMAN_REVIEW = "human_review"
    DENIED = "denied"


class DecisionMetrics:
    """Decision metrics for recording in spans."""
    
    def __init__(
        self,
        allowed: bool,
        requires_human_review: bool,
        confidence: float,
        risk_score: float,
        outcome: str
    ):
        self.allowed = allowed
        self.requires_human_review = requires_human_review
        self.confidence = confidence
        self.risk_score = risk_score
        self.outcome = outcome


class LatencyMetrics:
    """Latency metrics for recording in spans."""
    
    def __init__(
        self,
        total_ms: float,
        confidence_gate_ms: Optional[float] = None,
        policy_eval_ms: Optional[float] = None,
        risk_eval_ms: Optional[float] = None
    ):
        self.total_ms = total_ms
        self.confidence_gate_ms = confidence_gate_ms
        self.policy_eval_ms = policy_eval_ms
        self.risk_eval_ms = risk_eval_ms


class AgentTracer:
    """AI agent-aware tracer for enhanced observability."""
    
    def __init__(self, service_name: str = "lelu-python-sdk", service_version: str = "0.2.0") -> None:
        if OTEL_AVAILABLE:
            self.tracer = otel_trace.get_tracer(service_name, service_version)
        else:
            self.tracer = NoOpTracer()
    
    def start_agent_span(
        self,
        operation_name: str,
        agent_id: str,
        agent_type: str = AgentTypes.AUTONOMOUS,
        **attributes: Any
    ) -> Any:
        """Start a new span with AI agent semantic conventions."""
        span_attributes = {
            AIAgentAttributes.AGENT_ID: agent_id,
            AIAgentAttributes.AGENT_TYPE: agent_type,
            **attributes
        }
        
        span = self.tracer.start_span(operation_name)
        if hasattr(span, 'set_attributes'):
            span.set_attributes(span_attributes)
        
        return span
    
    def start_authorization_span(
        self,
        agent_id: str,
        action: str,
        confidence: float
    ) -> Any:
        """Start an authorization span."""
        return self.start_agent_span(
            "ai.agent.authorize",
            agent_id,
            AgentTypes.AUTONOMOUS,
            **{
                AIAgentAttributes.REQUEST_INTENT: action,
                AIAgentAttributes.REQUEST_CONFIDENCE: confidence,
                AIAgentAttributes.DECISION_TYPE: DecisionTypes.AUTONOMOUS,
            }
        )
    
    def start_delegation_span(
        self,
        delegator: str,
        delegatee: str
    ) -> Any:
        """Start a delegation span."""
        return self.start_agent_span(
            "ai.agent.delegate",
            delegator,
            AgentTypes.AUTONOMOUS,
            **{
                AIAgentAttributes.PARENT_AGENT: delegator,
                AIAgentAttributes.CHILD_AGENT: delegatee,
                AIAgentAttributes.DELEGATION_CHAIN: f"{delegator}→{delegatee}",
            }
        )
    
    def start_swarm_span(
        self,
        swarm_id: str,
        orchestrator: str,
        agents: List[str]
    ) -> Any:
        """Start a swarm orchestration span."""
        return self.start_agent_span(
            "ai.swarm.orchestration",
            orchestrator,
            agent_type="orchestrator",
            **{
                AIAgentAttributes.SWARM_ID: swarm_id,
                "ai.swarm.orchestrator": orchestrator,
                "ai.swarm.agents": agents,
                "ai.swarm.agent_count": len(agents),
            }
        )
    
    def record_policy_evaluation(
        self,
        span: Any,
        policy_name: str,
        policy_version: str,
        result: str,
        latency_ms: float
    ) -> None:
        """Record policy evaluation details in the span."""
        if not span or not hasattr(span, 'set_attributes'):
            return
        
        span.set_attributes({
            AIAgentAttributes.POLICY_NAME: policy_name,
            AIAgentAttributes.POLICY_VERSION: policy_version,
            AIAgentAttributes.POLICY_RESULT: result,
            AIAgentAttributes.POLICY_EVAL_MS: latency_ms,
        })
    
    def record_decision(
        self,
        span: Any,
        metrics: DecisionMetrics
    ) -> None:
        """Record the final authorization decision."""
        if not span or not hasattr(span, 'set_attributes'):
            return
        
        decision_type = DecisionTypes.AUTONOMOUS
        if metrics.requires_human_review:
            decision_type = DecisionTypes.HUMAN_REVIEW
        elif not metrics.allowed:
            decision_type = DecisionTypes.DENIED
        
        span.set_attributes({
            AIAgentAttributes.DECISION_TYPE: decision_type,
            AIAgentAttributes.DECISION_CONFIDENCE: metrics.confidence,
            AIAgentAttributes.DECISION_HUMAN_REVIEW: metrics.requires_human_review,
            AIAgentAttributes.DECISION_RISK_SCORE: metrics.risk_score,
            AIAgentAttributes.DECISION_OUTCOME: metrics.outcome,
        })
        
        # Set span status based on decision
        if hasattr(span, 'set_status'):
            if not metrics.allowed and not metrics.requires_human_review:
                span.set_status(OtelStatusCode.ERROR, "Authorization denied")
            else:
                span.set_status(OtelStatusCode.OK)
    
    def record_latency(
        self,
        span: Any,
        metrics: LatencyMetrics
    ) -> None:
        """Record operation latency."""
        if not span or not hasattr(span, 'set_attributes'):
            return
        
        attributes = {AIAgentAttributes.LATENCY_MS: metrics.total_ms}
        
        if metrics.confidence_gate_ms is not None:
            attributes[AIAgentAttributes.CONFIDENCE_GATE_MS] = metrics.confidence_gate_ms
        if metrics.policy_eval_ms is not None:
            attributes[AIAgentAttributes.POLICY_EVAL_MS] = metrics.policy_eval_ms
        if metrics.risk_eval_ms is not None:
            attributes[AIAgentAttributes.RISK_EVAL_MS] = metrics.risk_eval_ms
        
        span.set_attributes(attributes)
    
    def inject_correlation_context(
        self,
        span: Any,
        delegation_chain: str
    ) -> None:
        """Inject correlation context for multi-agent tracing."""
        if not span or not hasattr(span, 'set_attributes'):
            return
        
        span.set_attributes({
            AIAgentAttributes.DELEGATION_CHAIN: delegation_chain,
        })
    
    @contextmanager
    def agent_span(
        self,
        operation_name: str,
        agent_id: str,
        agent_type: str = AgentTypes.AUTONOMOUS,
        **attributes: Any
    ) -> Any:
        """Context manager for agent spans."""
        span = self.start_agent_span(operation_name, agent_id, agent_type, **attributes)
        
        try:
            yield span
            if hasattr(span, 'set_status'):
                span.set_status(OtelStatusCode.OK)
        except Exception as e:
            if hasattr(span, 'record_exception'):
                span.record_exception(e)
            if hasattr(span, 'set_status'):
                span.set_status(OtelStatusCode.ERROR, str(e))
            raise
        finally:
            if hasattr(span, 'end'):
                span.end()


# Default instance
agent_tracer = AgentTracer()


def get_agent_tracer(service_name: str = "lelu-python-sdk") -> AgentTracer:
    """Get an agent tracer instance."""
    return AgentTracer(service_name)